### PR TITLE
Entryfilelist parser

### PR DESCRIPTION
**Pull request by vswarte**

### CODE REVIEW

File: crates/formats/src/entryfilelist.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    #[error("Zlib error")]
+    Zlib,
+
+    #[error("Io error")]
+    Io(#[from] io::Error),
+}
+
+#[allow(unused)]
+pub struct EntryfilelistContainer<'a> {
```
Comment (garyttierney) [inline crates/formats/src/entryfilelist.rs:29]:
```suggestion
pub struct EntryFileListContainer<'a> {
```

File: crates/formats/src/entryfilelist.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            header: header.into_ref(),
+            compressed,
+        })
+    }
+
+    fn hint_size(&self) -> usize {
+        self.header.decompressed_size.get() as usize
+    }
+
+    pub fn decompress(&self) -> Result<Entryfilelist, EntryfilelistError> {
```
Comment (garyttierney) [inline crates/formats/src/entryfilelist.rs:53]: > Grabbing all the strings alone is pushing the 500mb rn and its still not done

Is that in reference to this?

File: crates/formats/src/io_ext/seek.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+use std::io::{self, ErrorKind, Seek, SeekFrom};
+
+pub trait SeekExt: Seek {
+    fn seek_until_alignment(&mut self, alignment: usize) -> io::Result<usize>;
+}
+
+impl<T: Seek> SeekExt for T {
```
Comment (garyttierney) [inline crates/formats/src/io_ext/seek.rs:7]: I'd make this work over `Read` and `std::io::copy(&mut read.take(required), &mut std::io::sink())` so we can use this on readers that don't support seeking.

File: crates/cli/src/bin/entryfilelist-test.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
+use std::{error::Error, path::PathBuf};
+
+use clap::Parser;
+use fstools_dvdbnd::{DvdBnd, FileKeyProvider};
+use fstools_formats::entryfilelist::EntryfilelistContainer;
+
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
```
Comment (garyttierney) [inline crates/cli/src/bin/entryfilelist-test.rs:8]: Could you move this to a `describe` command in the new CLI crate?

File: crates/formats/src/entryfilelist.rs (hunk 2 above)
Comment (vswarte) [reply crates/formats/src/entryfilelist.rs:53]: Nope that's the size of the text file generated by entryfilelist-test when piped to a file

File: crates/cli/src/bin/entryfilelist-test.rs (hunk 4 above)
Comment (vswarte) [reply crates/cli/src/bin/entryfilelist-test.rs]: Done per 3c9d315308f0fbc7d836d45e6535543734e5b6cf

File: crates/formats/src/entryfilelist.rs (hunk 1 above)
Comment (vswarte) [reply crates/formats/src/entryfilelist.rs]: Done per 3c9d315308f0fbc7d836d45e6535543734e5b6cf

File: crates/formats/src/io_ext/seek.rs (hunk 3 above)
Comment (vswarte) [reply crates/formats/src/io_ext/seek.rs]: [65a45f6](https://github.com/soulsmods/fstools-rs/pull/31/commits/65a45f6419e66c2e588cafebc48047ea7518d73c) removes it and replaces it with a call to read padding which I think is a bit better

File: crates/formats/src/entryfilelist.rs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    }
+}
+
+pub trait SectionElement: Sized {
+    fn read(reader: &mut impl Read) -> std::io::Result<Self>
+    where
+        Self: Sized;
+}
+
+impl<'a, TElement> Iterator for SectionIter<'a, TElement>
```
Comment (garyttierney) [inline crates/formats/src/entryfilelist.rs:149]: `T` is fine here, I think most people would figure out the `T` is the element produced by the iterator.

Comment (vswarte) [reply crates/formats/src/entryfilelist.rs]: Fixed per https://github.com/soulsmods/fstools-rs/pull/31/commits/99296a8272bd83e90e78bb5d0ae0d8c205146714

File: crates/formats/src/entryfilelist.rs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+impl<T, R: Read> Iterator for SectionIter<T, R>
+where
+    T: SectionElement,
+{
+    type Item = io::Result<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.entries_read < self.entry_count {
+            let result = (self.entries_read..self.entry_count)
```
Comment (garyttierney) [inline crates/formats/src/entryfilelist.rs:156]:
```rs
let result = (self.entries_read < self.entry_count).then(|| {
   self.entries_read += 1;
   T::read(&mut self.decoder)
});
```

**Review by garyttierney — APPROVED** [review]

**Pull request merged**